### PR TITLE
Refactor list and helptemplate printing to hod.table.

### LINKED ISSUE #98: hod list table text
Currently `hod list` prints label, jobid and then a job state. But the job state prints the job id again, followed by job state (Q - queued, R - running, etc) and then the hosts. The job state and hosts should be broken out into their own columns.

### PR DESCRIPTION
**Pull request by ehiggs**
This should fix #98.

### CODE REVIEW

File: hod/subcommands/helptemplate.py [hunk 1]
```diff
@@ -74,11 +75,11 @@ class HelpTemplateSubCommand(SubCommand):
 
     def run(self, args):
         """Run 'help-template' subcommand."""
+        os.environ['PBS_JOBID'] = '123.%s' % os.environ['PBS_DEFAULT']
```
Comment (boegel) [inline hod/subcommands/helptemplate.py]: say what? :)

File: hod/subcommands/helptemplate.py [hunk 2]
```diff
@@ -56,15 +58,14 @@ def mk_registry():
...
-    longest_value = max(value_names, key=len)
-    max_val_len = len(longest_value)
-    return '%%-%ds:\t%%-%ds\t%%s' % (max_name_len, max_val_len)
-
+def format_rows(fields, resolver):
+    """
+    Given a list of fields, return the template substituted versions of the text
+    """
+    rows = []
+    for v in sorted(fields.values(), key=lambda x: x.name):
```
Comment (boegel) [inline hod/subcommands/helptemplate.py]: no single-letter variables outside of list comprehensions please (the one in the lambda is OK too though)

File: hod/subcommands/listcmd.py [hunk 3]
```diff
@@ -55,14 +56,15 @@ def mk_fmt_str(labels, jobid, header):
     return '%%-%ds\t%%-%ds\t%%s' % (max_label_len, max_jobid_len)
```
Comment (boegel) [inline hod/subcommands/listcmd.py]: kill this function, no longer used?

File: hod/subcommands/listcmd.py [hunk 4]
```diff
@@ -55,14 +56,15 @@ def mk_fmt_str(labels, jobid, header):
...
 
 
-def format_cluster_info(cluster_info):
-    """Turn a list of 'ClusterInfo' objects into a 2-tuple of strings."""
+def format_rows(cluster_info):
```
Comment (boegel) [inline hod/subcommands/listcmd.py]: don't use the same name as the function in `helptemplate.py`, it will get very confusing, so rename them both (`format_list_rows` here?)

File: hod/subcommands/listcmd.py [hunk 5]
```diff
@@ -55,14 +56,15 @@ def mk_fmt_str(labels, jobid, header):
...
+    """Turn a list of 'ClusterInfo' objects into a list of strings."""
     ret = []
     for info in cluster_info:
         label = info.label if info.label is not None else '<no-label>'
         jobid = info.jobid
-        job = str(info.pbsjob) if info.pbsjob is not None else '<job-not-found>'
-        ret.append((label, jobid, job))
+        state = info.pbsjob.state if info.pbsjob is not None else '<job-not-found>'
+        hosts = info.pbsjob.hosts if info.pbsjob is not None else '<no-hosts-found>'
```
Comment (boegel) [inline hod/subcommands/listcmd.py]: maybe this is clearer?

```python
if info.pbsjob is not None:
    state, hosts = info.pbsjob.state, info.pbsjob.hosts
else:
    state, hosts = '<job-not-found>', '(none)'
```

File: hod/table.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        cols.append([row[col] for row in rows])
+    return cols
+
+def format_table(rows, headers=None):
+    """
+    Formats a table of data. 
+    Params
+    ------
+    rows - list of lists containing each row of data
+    headers - None or list of strings for the table header.
```
Comment (boegel) [inline hod/table.py]: use `@param rows`, @param headers`?

File: hod/table.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+    The number of headers must be the same as the number of columns or None.
+    """
+
+    if not rows:
+        raise ValueError('No rows to print')
+
+    if headers is not None and len(headers) != len(rows[0]):
```
Comment (boegel) [inline hod/table.py]: maybe also check whether all rows have same length?

`nub(map(len, rows)) == len(rows[0])`

`nub` is available in `vsc.utils.missing`

File: hod/table.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        out = fmt_str % tuple(headers)
+        out += '\n'
+    else:
+        cols = _pivot(rows)
+        fmt_str = _mk_fmt_str(cols)
+
+    for i, line in enumerate(rows):
+        out += fmt_str % tuple(line)
+        if i != len(rows)-1:
+            out += '\n'
```
Comment (boegel) [inline hod/table.py]: the `i != len` bit is a bit much, no?

```python
out += ''.join([fmt_str % tuple(line) for line in rows])
out += '\n'
```

File: hod/table.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+    if headers is not None and len(headers) != len(rows[0]):
+        raise ValueError('Headers and rows must have the same length')
+
+    out = ''
+    if headers:
+        cols = _pivot([headers] + rows)
+        fmt_str = _mk_fmt_str(cols)
+        out = fmt_str % tuple(headers)
+        out += '\n'
```
Comment (boegel) [inline hod/table.py]: collapse with previous line, and use `+=` on line above

```python
out += fmt_str % tuple(headers) + '\n'
```

File: hod/table.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    """Given a list of rows, pivot it into a list of columns - or vice versa."""
+    if not rows:
+        return []
+
+    cols = []
+    ncols = len(rows[0])
+    for col in range(ncols):
+        cols.append([row[col] for row in rows])
+    return cols
+
```
Comment (boegel) [inline hod/table.py]: two empty lines between top-level functions

File: hod/table.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+def _mk_fmt_str(fields):
+    """Calculate the format string for printing the template parameters."""
+    field_lens = []
+    for field in fields:
+        field_lens.append(max(map(len, field)))
+    fmt_str = ""
+    for i, field_len in enumerate(field_lens):
+        fmt_str += '%%-%ds' % field_len
+        if i != len(field_lens)-1:
+            fmt_str += '\t'
```
Comment (boegel) [inline hod/table.py]: same here:

```python
fmt_str = ''.join(['%%-%ds' % fl for fl in field_lens])
fmt_str += '\t'
```

or even

```python
fmt_str = ''.join(['%%-%ds' % fl for fl in field_lens]) + '\t' 
```

File: hod/table.py [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+@author: Ewan Higgs (Universiteit Gent)
+"""
+
+from itertools import chain
+
+def _mk_fmt_str(fields):
+    """Calculate the format string for printing the template parameters."""
+    field_lens = []
+    for field in fields:
+        field_lens.append(max(map(len, field)))
```
Comment (boegel) [inline hod/table.py]:
```python
field_lens = [max(map(len, field)) for field in fields]
```

File: hod/table.py [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+# You should have received a copy of the GNU General Public License
+# along with hanythingondemand. If not, see <http://www.gnu.org/licenses/>.
+# #
+"""
+Utility functions for printing text in tables.
+
+@author: Ewan Higgs (Universiteit Gent)
+"""
+
+from itertools import chain
```
Comment (boegel) [inline hod/table.py]: no used, please drop

File: test/unit/subcommands/test_subcommands_helptemplate.py [hunk 14]
```diff
@@ -36,8 +37,11 @@
 class TestHelpTemplateApplication(unittest.TestCase):
     def test_run(self):
         app = HelpTemplateSubCommand()
-        with patch('hod.config.template.mklocalworkdir', return_value='localworkdir'):
-            app.run([])
+        environ = dict(PBS_DEFAULT='pbsmaster')
+        environ.update(os.environ)
```
Comment (boegel) [inline test/unit/subcommands/test_subcommands_helptemplate.py]: what if `$PBS_DEFAULT` is defined in the environment?

```python
test_environ = copy.deepcopy(os.environ)
test_environ['PBS_DEFAULT'] = 'pbsmaster'
with patch('os.environ', test_environ):
```

**Comment by boegel** [comment]
@ehiggs: reviewed, ping me when you're done with the remarks

File: hod/subcommands/helptemplate.py (hunk 1 above)
Comment (ehiggs) [inline hod/subcommands/helptemplate.py]: Someone deep down wants `PBS_JOBID` defined. So I define it.

File: test/unit/subcommands/test_subcommands_helptemplate.py (hunk 14 above)
Comment (ehiggs) [inline test/unit/subcommands/test_subcommands_helptemplate.py]: That's fine. It needs to be defined.

File: hod/table.py (hunk 6 above)
Comment (ehiggs) [inline hod/table.py]: That's javadoc notation. `reST` uses `:param` but we don't even use that.  What I have is somewhat similar to the [Google styleguide suggestion](https://google-styleguide.googlecode.com/svn/trunk/pyguide.html) but the docstrings aren't really consistent across the project (it's an end user tool, not a library).

File: hod/table.py (hunk 7 above)
Comment (ehiggs) [inline hod/table.py]: This has been addressed.

File: hod/subcommands/helptemplate.py [hunk 15]
```diff
@@ -74,11 +75,15 @@ class HelpTemplateSubCommand(SubCommand):
 
     def run(self, args):
         """Run 'help-template' subcommand."""
-        optparser = HelpTemplateOptions(go_args=args, envvar_prefix=self.envvar_prefix, usage=self.usage_txt)
+        # We need PBS_JOBID defined in the environment to generate localworkdirs
+        if 'PBS_DEFAULT' not in os.environ:
+            os.environ['PBS_DEFAULT'] = 'pbs-master'
+        os.environ['PBS_JOBID'] = '123.%s' % os.environ['PBS_DEFAULT']
```
Comment (boegel) [inline hod/subcommands/helptemplate.py]: this is just wrong, especially without any form of logging...

where is `$PBS_JOBID` required exactly? deal with not having it defined there somehow, by using a fallback value if `$PBS_JOBID` is not defined, not here, this is a hack at best

File: hod/table.py (hunk 7 above)
Comment (boegel) [inline hod/table.py]: ok

File: test/unit/subcommands/test_subcommands_helptemplate.py [hunk 16]
```diff
@@ -36,8 +37,11 @@
 class TestHelpTemplateApplication(unittest.TestCase):
     def test_run(self):
         app = HelpTemplateSubCommand()
-        with patch('hod.config.template.mklocalworkdir', return_value='localworkdir'):
-            app.run([])
+        test_environ = dict(PBS_DEFAULT='pbsmaster')
```
Comment (boegel) [inline test/unit/subcommands/test_subcommands_helptemplate.py]: why is this suddenly required? is this related to the `$PBS_JOBID` hack?

**Comment by boegel** [comment]
@ehiggs: rereviewed, the `$PBS_JOBID` hack is a blocker imho

File: test/unit/subcommands/test_subcommands_helptemplate.py (hunk 16 above)
Comment (ehiggs) [inline test/unit/subcommands/test_subcommands_helptemplate.py]: Yes.

Comment (boegel) [inline test/unit/subcommands/test_subcommands_helptemplate.py]: OK, then I'd like to see it disappear too. ;-)

File: hod/subcommands/helptemplate.py (hunk 15 above)
Comment (ehiggs) [inline hod/subcommands/helptemplate.py]: Something is required because the help-template command needs to print the localworkdir. The localworkdir now uses the jobid to group the underlying local work directories. But in the case of printing the job it, we don't have a job and hence no jobid.

Comment (boegel) [inline hod/subcommands/helptemplate.py]: OK, but then handle that where you need the job ID somehow, not by injecting a dummy value in the environment.

What if there are other places (in the future) that need to have/print the localworkdir without being in a job?

Comment (ehiggs) [inline hod/subcommands/helptemplate.py]: This is a special command for printing diagnostic information outside a job. In all other cases, it should be an error to try to figure out a localworkdir without a job id. There could be some refactoring that takes place, but that would just be to pull it from the environment and then thread the job id to all calls until it gets to the localworkdir.

File: test/unit/subcommands/test_subcommands_helptemplate.py (hunk 16 above)
Comment (ehiggs) [inline test/unit/subcommands/test_subcommands_helptemplate.py]: Me too. But it's running outside the normal job context.

**Comment by boegel** [comment]
Still feel a bit queasy about the `$PBS_JOBID` hack, but at least it's implemented cleanly now.

Going in, thanks @ehiggs!

**Comment by ehiggs** [comment]
Thanks!
